{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "formed-marshall",
   "metadata": {},
   "source": [
    "# Introduction to Whole Exome Sequencing Variant Calling\n",
    "\n",
    "This tutorial will provide an example pipeline for calling variants in whole exome sequencing (WES) data across multiple samples, assuming the samples are paired-end. It is strongly recommended that these step be run on a large compute resource such a high-performance compute cluster (HPC) or a cloud resource.  For HPC, these commands should be wrapped inside a shell script with the HPC scheduling language installed on the cluster.  I will try to denote which ones are critical for the cluster, versus which ones can be run locally.\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:2px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressed-corpus",
   "metadata": {},
   "source": [
    "## STEP0: Fastq Raw Quality Check\n",
    "\n",
    "Before any sample processing can begin, it is critical to know the initial state of the raw sequences. This is typically delivered from the sequencer in fastq format (.fastq/.fq) or gzipped (fastq.gz/fq.gz) so the data is compressed to decrease file size and storage use. \n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)|\n",
    "| Dependencies             | Perl, Java + JDK |\n",
    "| Tested on                | FastQC==v0.11.9, Ubuntu 18.04.4 LTS (Bionic Beaver)| \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | html report and compressed images |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "### Commands\n",
    "For <span style=\"background:yellow;color:black\">**_each fastq file_**</span>, you will want to run the following command:\n",
    "<br/>\n",
    "\n",
    "<p style=\"background:white\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">fastqc myFile.fastq.gz -o /path/to/my/output/directory\n",
    "</code>\n",
    "</p>\n",
    "    \n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt; insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS) &gt; <br/>\n",
    "    # dependency of fastqc\n",
    "    export PERL5LIB=/path/to/perl <br/>\n",
    "    # declare and define variables\n",
    "    fastqc=\"/path/to/fastqc\"\n",
    "    listOfFastqs=\"/path/to/file/with/fastq/pathnames.txt\"\n",
    "    outDir=\"/path/to/output/directory/\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP0_fastqc.parallel.log --resume\" <br/>  \n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles &lt; \\$listOfFastqs <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $fastqc {} -o $outDir\" ::: ${allFiles[@]} <br/>\n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "\n",
    "For each fastq file an html report and a zipped directory will be produced.  The zipped directory contains the same information ast the html report; the difference being that the zipped directory contains all the report images as image files that can be embedded into documents/presentations/etc...\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [ ] <span style=\"color:blue\">**Sequence trimming and clean up**</span>\n",
    "- [ ] Fastq Check on Trimmed Fastq\n",
    "- [ ] Alignment/Mapping\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>FastQC</b><br/>\n",
    "* There is not a formal paper publication on this software, so cite the software as open-source code or webpage.<br/>\n",
    "* For <a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/\">FastQC Docs</a> click on the link.\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contrary-referral",
   "metadata": {},
   "source": [
    "## STEP1: Basic quality trimming and adapter removal\n",
    "\n",
    "Based on the output of fastqc, you will want to make sure poor quality bases are trimmed, any adapter contamination as a result of the library prep or sequencing linkers are removed, and any paired reads that may be too short are *removed together.  This is critical for paired-end reads as the order between the forward (R1) and the reverse (R2) must be maintained.*  Reads that are too short will ineviitbaly lead to high multimappers in the alignment phase and can computationally slow down the aligner.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [cutadapt](https://cutadapt.readthedocs.io/en/stable/)|\n",
    "| Dependencies             | Python3 |\n",
    "| Tested on                | cutadapt==v3.2, Python>=3.6.9, Ubuntu 18.04.4 LTS (Bionic Beaver)| \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "Given cutadapt is a python package, it can be installed using pip:\n",
    "\n",
    "```\n",
    "    pip3 install --user cutadapt==3.2\n",
    "```\n",
    "\n",
    "Assuming FastQC shows Illumina Universal Adapter contamination, we will want to remove those adapter sequences from both reads ends, trim/remove poor quality sequences, and dicard read pairs that may be too short.  One thing to consider, if the sequences were run on the NovaSeq, quality scores are no longer granular!  The NovaSeq bins Phred quality scores to 4 potential bins of either 2, 12, 23, and 37.  Therfore, when selecting a quality cut-off know that if you pick something between the bin values, it will ony remove/trim sequences based on the lower bin.  [Heng Li has great blog](https://lh3.github.io/2017/07/24/on-nonvaseq-base-quality) about this if you are interested.\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Commands\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each pair of fastq files_**</span>, you will want to run the following command:\n",
    "\n",
    "<p style=\"background:white\">\n",
    "<code style=\"background:grey;color:black;font-size:16px;margin-left:40px\">cutadapt -a AGATCGGAAGAG -A AGATCGGAAGAG -q 30 --minimum-length=10 --pair-filter=any -o Read1Cleaned_1.fq.gz -p Read2Cleaned_2.fq.gz Read1.fq.gz Read2.fq.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    export PERL5LIB=/path/to/perl\n",
    "    export PYTHONPATH=/path/to/python/site-packages <br/>\n",
    "    # declare and define variables\n",
    "    cutadapt=\"/path/to/cutadapt\"\n",
    "    listOfPEfastqs=\"/path/to/fastqPE/prefixs.txt\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP1_cutadapt.parallel.log --resume\" <br/>\n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles &lt; \\$listOfPEfastqs <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $cutadapt -a AGATCGGAAGAG -A AGATCGGAAGAG -q 30 --minimum-length=10 --pair-filter=any -o {}Cleaned_1.fq.gz -p {}Cleaned_2.fq.gz {}1.fq.gz {}2.fq.gz\" ::: ${allFiles[@]} <br/>\n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [ ] <span style=\"color:blue\">**Fastq Check on Trimmed Fastq**</span>\n",
    "- [ ] Alignment/Mapping\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>cutadapt Hallmark Paper</b><br/>\n",
    "* <i>Martin, M. (2011). Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal,    17(1), pp. 10-12. <a href=\"https://doi.org/10.14806/ej.17.1.200\">doi:https://doi.org/10.14806/ej.17.1.200</a></i> <br/>\n",
    "* For <a href=\"https://cutadapt.readthedocs.io/en/stable/guide.html\">cutadapt Docs</a> click on the link.\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-expert",
   "metadata": {},
   "source": [
    "<br/><br/>\n",
    "\n",
    "## STEP2: Fastq Cleaned Quality Check\n",
    "\n",
    "This is the exact same step as STEP0, with the exception that now you are feeding in the the cleaned fastq files generated in STEP1 that have now been cleaned.  This step is to ensure all cleaning was performed as expected.\n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)|\n",
    "| Dependencies             | Perl, Java + JDK |\n",
    "| Tested on                | FastQC==v0.11.9, Ubuntu 18.04.4 LTS (Bionic Beaver) | \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | html report and compressed images |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "### Commands\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each fastq file_**</span>, you will want to run the following command:\n",
    "\n",
    "<p style=\"background:white\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">fastqc myCleanedSTEP1File.fastq.gz -o /path/to/my/output/directory/\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    # dependency of fastqc\n",
    "    export PERL5LIB=/path/to/perl <br/>\n",
    "    # declare and define variables\n",
    "    fastqc=\"/path/to/fastqc\"\n",
    "    listOfFastqs=\"/path/to/file/with/fastq/cleaned/pathnames.txt\"\n",
    "    outDir=\"/path/to/output/directory/\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP2_fastqc.parallel.log --resume\" <br/>\n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles &lt; \\$listOfFastqs <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $fastqc {} -o $outDir\" ::: ${allFiles[@]} <br/>\n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [ ] <span style=\"color:blue\">**Alignment/Mapping**</span>\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>FastQC</b><br/>\n",
    "* There is not a formal paper publication on this software, so cite the software as open-source code or webpage.<br/>\n",
    "* For <a href=\"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/\">FastQC Docs</a> click on the link.\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "speaking-engineer",
   "metadata": {},
   "source": [
    "## STEP3: Aligning Reads to Reference Genome\n",
    "\n",
    "This is typically the most computationally expensive part of the analysis and should be performed on a machine with large enough RAM/memory and cores.  Since this is DNA, there is no need to use a splice-aware aligner, therefore a DNA aligner such as bowtie2 or bwa is fine.  \n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [BWA](http://bio-bwa.sourceforge.net/), [Samtools](http://www.htslib.org/) |\n",
    "| Dependencies             |  |\n",
    "| Tested on                | bwa==v0.7.11, samtools==v1.8 | \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Ouput File Type          | alignment file w/extension *.sam or *.bam |\n",
    "\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Commands\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each pair of fastq files_**</span>, you will want to run the following command.  This assumes you have 12 cores available for threading -- 6 for alignment and 6 for sorting: \n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bwa mem -t 6 /path/to/myBWAIndex/filePrefix Read1Cleaned_1.fq.gz Read2Cleaned_1.fq.gz | samtools sort -@6 -o SampleID_trimmed_sorted.bam -\n",
    "</code> \n",
    "</p>\n",
    "<br/>\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Tip:</b> In order to use BWA, you must index your reference genome first.  This only needs to be done one time per reference genome.  This can be accomlished by using <code style=\"background:black;color:white\">bwa index youReferenceGenome.fasta</code>.  For more information, <a href=\"http://bio-bwa.sourceforge.net/bwa.shtml\">look at the bwa documentation here</a>.</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt;<br/>\n",
    "    module load samtools/1.8 # or declare variable to samtools executable <br/>\n",
    "    # declare and define variables\n",
    "    bwa=\"/path/to/bwa\"\n",
    "    bwaIndexGenome=\"/path/to/bwa_indices/including/prefix\"\n",
    "    listOfPEfastqs=\"/path/to/fastqPE/prefixs.txt\"\n",
    "    #samtools=\"/path/to/samtools/if/needed\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 50G --delay 0.2 --jobs 2 -u --progress --joblog STEP3_WES_bwa_alignment.parallel.log --resume\" <br/>\n",
    "    # store the prefix file list as an array\n",
    "    readarray allFiles &lt; \\$listOfPEfastqs <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $bwa mem -t 6 $bwaIndexGenome {}R1.fq.gz {}R2.fq.gz | samtools sort -@6 -o {}trimmed_sorted.bam -\" ::: ${allFiles[@]} <br/>\n",
    "    \n",
    "</code>\n",
    "</p>\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [ ] <span style=\"color:blue\">**Alignment Quality Check**</span>\n",
    "- [ ] <span style=\"color:blue\">**Duplicate Removal**</span>\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>Burrows-Wheeler Alignment Tool (BWA) Hallmark Paper</b><br/>\n",
    "* <i>Li, H., & Durbin, R. (2009). Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics (Oxford, England), 25(14), 1754–1760. <a href=\"https://doi.org/10.1093/bioinformatics/btp324\">https://doi.org/10.1093/bioinformatics/btp324</a>.</i><br/>\n",
    "* For <a href=\"http://bio-bwa.sourceforge.net/\">BWA Docs</a> click on the link.\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "solved-firmware",
   "metadata": {},
   "source": [
    "## STEP4: Collect and Assess Alignment Quality + Remove Duplicates\n",
    "\n",
    "This can easily be performed on a local computer as the memory and storage footprint are relatively low.  This collect alignment information and statistics, insert size metrics, and duplicate metrics.  The duplicate read removal step is critical for WES and variant calling.  In this case, I physically remove the duplicate reads instead of \"marking\" the duplicates, in case the GATK pipeline is not used for variant calling.  All of these metrics collected should be looked at and if there are enough samples, graphically reprensented using your own custom code to ensure the differences can be explained within and between samples.  **It may be the case that after looking at these metrics, alignmnet paramters need to be adjusted and STEP3 needs to be re-run and re-assessed to fit the needs of the project.**\n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | Local, HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [Picard](https://broadinstitute.github.io/picard/), [Samtools](http://www.htslib.org/) |\n",
    "| Dependencies             | Java + JDK |\n",
    "| Tested on                | Picard==v2.21.1, samtools==v1.8 | \n",
    "| Input File Type          | alignment file w/extension *.bam or *sam |\n",
    "| Output File Type         | series of txt files and alignment file w/extension *.bam and index *.bai | \n",
    "\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Commands\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each alignment file_**</span>, you will want to run the following sets of commands:\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">samtools flagstat alignmentFileName.bam > alignmentFileName.flagstat\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar CollectAlignmentSummaryMetrics I=alignmentFileName.bam O=picard.metrics.txt R=refGenome.fasta\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar CollectInsertSizeMetrics I=alignmentFileName.bam O=insert_size_metrics.txt H=insert_size_histogram.pdf\n",
    "</code> \n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar MarkDuplicates I=alignmentFileName.bam O=alignmentNoDups.bam M=markDupMetrics.txt REMOVE_DUPLICATES=true\n",
    "</code>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">samtools index alignmentNoDups.bam\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available alignment files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    module load java\n",
    "    module load picard/2.21.1\n",
    "    module load samtools/1.8\n",
    "    export PERL5LIB=/home/bin/perl <br/>\n",
    "    # declare and define variables\n",
    "    #picardJar=\"/path/to/picard/if/needed\"\n",
    "    #samtools=\"/path/to/samtools/if/needed\"\n",
    "    bamPath=\"/path/to/bam/output/location/directory/\"\n",
    "    outDir=\"/path/to/output/results/\"\n",
    "    bamList=\"/path/to/file/with/bamPrefix.txt\"\n",
    "    fastaFile=\"/path/to/reference/genome.fasta\" <br/>\n",
    "    # store the prefix file list as an array\n",
    "    readarray allBamFiles &lt; $bamList <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    echo \"STEP4: 1/5...flagstats\"\n",
    "    for bamFile in ${bamPath}*.bam;\n",
    "    do\n",
    "      time samtools flagstat ${bamFile} &gt; ${bamFile}.flagstat\n",
    "    done\n",
    "    wait\n",
    "    sleep 60 <br/>                         \n",
    "    echo \"STEP4: 2/5...picard alignment metrics\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_alignmentMetrics_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar CollectAlignmentSummaryMetrics I=${bamPath}{}.bam O=${outDir}{}.picard.metrics.txt R=$fastaFile\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60 <br/>\n",
    "    echo \"STEP4: 3/5...collect insert size metrics\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_collectInsertMetrics_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar CollectInsertSizeMetrics I=${bamPath}{}.bam O=${outDir}{}insert_size_metrics.txt H=${outDir}{}insert_size_histogram.pdf\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60 <br/>\n",
    "    echo \"STEP4: 4/5...mark duplicates\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_markDuplicates_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar MarkDuplicates I=${bamPath}{}.bam O=${outDir}{}noDups.bam M=${outDir}{}markDupMetrics.txt REMOVE_DUPLICATES=true\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60 <br/>\n",
    "    echo \"STEP4: 5/5...index new bam\"\n",
    "    for newBam in ${outDir}*.bam;\n",
    "    do\n",
    "        time samtools index ${newBam}\n",
    "    done\n",
    "    wait\n",
    "    sleep 60 <br/>\n",
    "                                                                     \n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [ ] <span style=\"color:blue\">**Independent Variant Calling**</span>\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>SamTools Hallmark Paper</b><br/>\n",
    "* <i>Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G., Durbin, R., & 1000 Genome Project Data Processing Subgroup (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics (Oxford, England), 25(16), 2078–2079. <a href=\"https://doi.org/10.1093/bioinformatics/btp352\">https://doi.org/10.1093/bioinformatics/btp352<a/>.</i><br/>\n",
    "* For <a href=\"http://www.htslib.org/doc/samtools.html\">SamTools Docs</a> click on the link.<br/>\n",
    "* <a href=\"https://davetang.org/wiki/tiki-index.php?page=SAMTools\">Dave Tang's Blog</a> also has a great wiki and tutorial for SamTools <br/><br/>\n",
    "\n",
    "<b>Picard Tools</b><br/>\n",
    "    * For <a href=\"https://broadinstitute.github.io/picard/\">Picard Docs</a> click on the link.<br/>\n",
    "    * There does not seem to be a formal publication, so please cite their GitHub page located <a href=\"https://github.com/broadinstitute/picard\">here</a><br/>\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pharmaceutical-tracker",
   "metadata": {},
   "source": [
    "## STEP5: Independent Variant Calling per Sample\n",
    "\n",
    "After alignment is completed and QC is acceptable, then it is ready for variant calling.  When picking a variant caller is is important to understand the type of samples and research question.  For example, if a sample is expected to be a tumor sample with lots of copy number alterations or a non-diploid organism, algorithms such as deepVariant will not be suitable since it assumes diploid chromosomes learned from the model genome.  Additionally, if one is working with matched tumor/normal pairs, it is best to select a variant caller suitable for that purpose and one may even consider generating a panel of normals for this particular purpose.  For independent sample calling against a reference genome, that may be sensitive to ploidy issues or tumor only samples (i.e. expectation of CNVs), I use Platypus.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [Platypus](https://www.well.ox.ac.uk/research/research-groups/lunter-group/lunter-group/platypus-a-haplotype-based-variant-caller-for-next-generation-sequence-data) |\n",
    "| Dependencies             | Python2.7 |\n",
    "| Tested on                | Platypus==v0.8.1, Python==v2.7.6 | \n",
    "| Input File Type          | alignment file w/extension *.bam or *sam |\n",
    "| Output File Type         | variant call format file w/extension .vcf.gz | \n",
    "\n",
    "\n",
    "### Commands\n",
    "For <span style=\"background:yellow;color:black\">**_each alignment file_**</span>, you will want to run the following sets of commands:<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px;\">python Platypus.sh callVariants --bamFiles=noDups.bam --refFile=yourRefGenome.fasta --output=noDups.vcf --nCPU=5\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available bam alignment files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    platypus=\"/path/to/Platypus.py\"\n",
    "    inputBamsFilePrefix=\"/path/to/bamPrefixNames.txt\"\n",
    "    referenceFasta=\"/path/to/reference/genome.fasta\"\n",
    "    inputDir=\"/path/to/directory/of/alignment/bam/files/\" <br/>\n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles &lt; \\$inputBamsFilePrefix <br/>\n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    gnuParallel=\"/path/to/gnu/parallel --tag --memfree 20G --delay 0.2 --jobs 4 -u --progress --joblog STEP5.parallel.log --resume\" <br/>\n",
    "    $gnuParallel \"time python $platypus callVariants --bamFiles=$inputDir{}.bam --refFile=$referenceFasta --output=$outDir{}.vcf --nCPU=5\" ::: ${allFiles[@]} <br/>\n",
    "    \n",
    "    \n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [ ] <span style=\"color:blue\">**Independent Variant Filtering**</span>\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split multiallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>Platypus Hallmark Paper</b><br/>\n",
    "* <i>Andy Rimmer, Hang Phan, Iain Mathieson, Zamin Iqbal, Stephen R. F. Twigg, WGS500 Consortium, Andrew O. M. Wilkie, Gil McVean, Gerton Lunter. Integrating mapping-, assembly- and haplotype-based approaches for calling variants in clinical sequencing applications. Nature Genetics (2014) <a href=\"https://doi.org/10.1038/ng.3036\">doi:10.1038/ng.3036</a></i><br/>\n",
    "* For <a href=\"https://www.rdm.ox.ac.uk/research/lunter-group/lunter-group/platypus-documentation\">Platypus Docs</a> click on the link. <br/><br/>\n",
    "<b>VCF format</b><br/>\n",
    "* For infomration regarding <a href=\"https://samtools.github.io/hts-specs/VCFv4.2.pdf\">VCF formatting</a>, please click on the link.  As a warning, please double check the VCF file version you are using.  This can be done by looking at the first few lines in the header.  The format slighlty changes from version to version.\n",
    "</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specific-trading",
   "metadata": {},
   "source": [
    "## STEP7: Create Multi-Sample VCF & Joint/Backfill/Square Variant Calling\n",
    "This step merges all the filter individual VCFs generated into one file.  You will notice, that since the variants are not called in all samples, that there are some where genotypes are missing, denoted ./. or a .  This means a variant was not called usually because there are no alternate alleles at that position and therefore reference only, or the variant did not pass certain quality control filters.  In order to force call these, we use Platypus to fill in those missing genotypes and allele information.  This allows us to be able to compare samples as that variant using a Fisher's Exact test to determine if two samples are statistically significant based on the variant allele frequency.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | Local, HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [BCFtools](https://samtools.github.io/bcftools/), [htslib](https://github.com/samtools/htslib/releases/tag/1.11), [Platypus](https://www.well.ox.ac.uk/research/research-groups/lunter-group/lunter-group/platypus-a-haplotype-based-variant-caller-for-next-generation-sequence-data) |\n",
    "| Dependencies             | python2.7 |\n",
    "| Tested on                | BCFtools==v1.11, htslib==v1.11, Platypus==v0.8.1, Python==v2.7.6 | \n",
    "| Input File Type          | vcf files |\n",
    "| Output File Type         | variant call format file w/extension .vcf.gz and index .tbi | \n",
    "\n",
    "\n",
    "### Commands\n",
    "Call this command <span style=\"background:yellow;color:black\">**one time**</span> no matter how many vcf files are used!<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px;\">bcftools merge -Ov -o myMergedVcf.vcf noDups1.vcf.gz noDups2.vcf.gz noDups3.vcf.gz &lt;insert as many vcfs as you have&gt; \n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bgzip myMergedVcf.vcf</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">tabix myMergedVcf.vcf.gz</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">python Platypus.sh --bamFiles=noDup1.bam,noDup2.bam,noDup3.bam --refFile=yourRefGenome.fasta --output=myMergedVcfBackfilled.vcf --minPosterior=0 --getVariantsFromBam=0 --source=myMergedVcf.vcf.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bgzip myMergedVcfBackfilled.vcf</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">tabix myMergedVcfBackfilled.vcf.gz</code> \n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler_**</span>:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    bcftools=\"/path/to/bcftools/executable\"\n",
    "    tabix=\"/path/to/tabix/executable\"\n",
    "    bgzip=\"/path/to/bgzip/executable\"\n",
    "    platypus=\"/path/to/Platypus.py\"\n",
    "    python=\"/path/to/python/executable\"\n",
    "    referenceFasta=\"/path/to/reference/genome.fasta\"\n",
    "    inputDir=\"/path/to/directory/of/alignment/bam/files/\"\n",
    "    outputDir=\"/path/to/output/directory/ <br/>\n",
    "    $bcftools merge -Ov -o ${outputDir}myMergedVcf.vcf ${inputDir}noDups1.vcf.gz ${inputDir}noDups2.vcf.gz ${inputDir}noDups3.vcf.gz\n",
    "    $bgzip ${outputDir}myMergedVcf.vcf\n",
    "    $tabix ${outputDir}myMergedVcf.vcf.gz\n",
    "    $python $platypus --bamFiles=${inputDir}noDup1.bam,${inputDir}noDup2.bam,${inputDir}noDup3.bam --refFile=${referenceFasta} --output=${outputDir}myMergedVcfBackfilled.vcf --minPosterior=0 --getVariantsFromBam=0 --source=${inputDir}myMergedVcf.vcf.gz\n",
    "    $bgzip ${outputDir}myMergedVcfBackfilled.vcf \n",
    "    $tabix ${outputDir}myMergedVcfBackfilled.vcf.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [x] Independent Variant Filtering\n",
    "- [x] Multi-sample VCF Generation\n",
    "- [x] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] <span style=\"color:blue\">**Split multiallelic variants**</span>\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>Platypus Hallmark Paper</b><br/>\n",
    "* <i>Andy Rimmer, Hang Phan, Iain Mathieson, Zamin Iqbal, Stephen R. F. Twigg, WGS500 Consortium, Andrew O. M. Wilkie, Gil McVean, Gerton Lunter. Integrating mapping-, assembly- and haplotype-based approaches for calling variants in clinical sequencing applications. Nature Genetics (2014) <a href=\"https://doi.org/10.1038/ng.3036\">doi:10.1038/ng.3036</a></i><br/>\n",
    "* For <a href=\"https://www.rdm.ox.ac.uk/research/lunter-group/lunter-group/platypus-documentation\">Platypus Docs</a> click on the link. <br/><br/>\n",
    "<b>VCF format</b><br/>\n",
    "* For infomration regarding <a href=\"https://samtools.github.io/hts-specs/VCFv4.2.pdf\">VCF formatting</a>, please click on the link.  As a warning, please double check the VCF file version you are using.  This can be done by looking at the first few lines in the header.  The format slighlty changes from version to version. <br/><br/>\n",
    "<b>BCFtools Hallmark Papers</b><br/>\n",
    "* <i> Li H. (2011). A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. Bioinformatics (Oxford, England), 27(21), 2987–2993. <a href=\"https://doi.org/10.1093/bioinformatics/btr509\">https://doi.org/10.1093/bioinformatics/btr509</a></i><br/>\n",
    "* <i>Danecek, P., & McCarthy, S. A. (2017). BCFtools/csq: haplotype-aware variant consequences. Bioinformatics (Oxford, England), 33(13), 2037–2039. <a href=\"https://doi.org/10.1093/bioinformatics/btx100\">https://doi.org/10.1093/bioinformatics/btx100</a></i><br/>\n",
    "* For <a href=\"https://samtools.github.io/bcftools/bcftools.html\">BCFtools Docs</a>, click on the link. <br/><br/>\n",
    "</div>\n",
    "\n",
    "<br/>\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "laden-intervention",
   "metadata": {},
   "source": [
    "## STEP8: Split Multi-Allelic Variants\n",
    "Some times, variants are not biallelic.  In this case, the default for bctools and platypus is to comma-separate the allele on the same variant position.  This can cause issues with annotation and filtering since the ref/alt will not match unless the same multi-allelic variants exist in the database, which is not likely for most databases.  Therefore, we split the multi-allelic variants to have the same position span multiple, independent variant lines in the vcf file to force each position to be biallelic.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | Local, HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [BCFtools](https://samtools.github.io/bcftools/), [htslib](https://github.com/samtools/htslib/releases/tag/1.11) |\n",
    "| Dependencies             |  |\n",
    "| Tested on                | BCFtools==v1.11, htslib==v1.11 | \n",
    "| Input File Type          | vcf file |\n",
    "| Output File Type         | variant call format file w/extension .vcf.gz and index .tbi | \n",
    "\n",
    "\n",
    "### Commands\n",
    "Call each of these commands <span style=\"background:yellow;color:black\">**one time**</span> and sequentially:\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bcftools norm -m -both --do-not-normalize -Ov -o  myMergedVcfBiallelic.vcf myMergedVcfBackfilled.vcf.gz</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bgzip myMergedVcfBiallelic.vcf</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">tabix myMergedVcfBiallelic.vcf.gz</code> \n",
    "</p>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler_**</span>:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    bcftools=\"/path/to/bcftools/executable\"\n",
    "    tabix=\"/path/to/tabix/executable\"\n",
    "    bgzip=\"/path/to/bgzip/executable\"\n",
    "    outDir=\"/path/to/output/directory/\" <br/>\n",
    "    $bcftools norm -m -both --do-not-normalize -Ov -o ${outDir}myMergedVcfBiallelic.vcf myMergedVcfBackfilled.vcf.gz <br/>\n",
    "    $bgzip ${outDir}myMergedVcfBiallelic.vcf <br/>\n",
    "    $tabix ${outDir}myMergedVcfBiallelic.vcf.gz <br/>\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [x] Independent Variant Filtering\n",
    "- [x] Multi-sample VCF Generation\n",
    "- [x] Joint Variant Calling/Backfill/Squaring\n",
    "- [x] Split multiallelic variants\n",
    "- [ ] <span style=\"color:blue\">**Joint Variant Filtering**</span>\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "possible-bathroom",
   "metadata": {},
   "source": [
    "## STEP9: Joint Variant Filtering and QC Check\n",
    "Backfilling the variant calls may produce small amount of poor variants, i.e variants that perform poor across most samples.  These should be filtered.  It is also worth checking the breakdown of your filtering before you apply the filter.  Depending on how many variants are effected, you may want to be less stringent on filtering.  Also, you will notice a QC check using BCFtools stats command. A good quality check here is to look at the transition/transversion ratio (ts/tv).  The ts/tv ratio in WGS diploid humans is typically around 2-2.1, with WES being higher around 2.8, due to bias in increased coding regions.  See resources section below for more information.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | Local, HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [BCFtools](https://samtools.github.io/bcftools/), [htslib](https://github.com/samtools/htslib/releases/tag/1.11) |\n",
    "| Dependencies             |  |\n",
    "| Tested on                | BCFtools==v1.11, htslib==v1.11 | \n",
    "| Input File Type          | vcf file |\n",
    "| Output File Type         | variant call format file w/extension .vcf.gz and index .tbi | \n",
    "\n",
    "\n",
    "### Commands\n",
    "Call each of these commands <span style=\"background:yellow;color:black\">**one time**</span> and sequentially:\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bcftools filter -i 'FILTER=\"PASS\"' -Ov -o myMergedVcfBiallelicFiltered.vcf myMergedVcfBiallelic.vcf.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bgzip myMergedVcfBiallelicFiltered.vcf\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">tabix myMergedVcfBiallelicFiltered.vcf.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bcftools stats myMergedVcfBiallelicFiltered.vcf.gz > statistics.txt\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler_**</span>:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &lt;insert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt; <br/>\n",
    "    bcftools=\"/path/to/bcftools/executable\"\n",
    "    tabix=\"/path/to/tabix/executable\"\n",
    "    bgzip=\"/path/to/bgzip/executable\"\n",
    "    outDir=\"/path/to/output/directory/\" <br/>\n",
    "    $bcftools filter -i 'FILTER=\"PASS\"' -Ov -o ${outDir}myMergedVcfBiallelicFiltered.vcf <br/>\n",
    "    $bgzip ${outDir}myMergedVcfBiallelicFiltered.vcf <br/>\n",
    "    $tabix ${outDir}myMergedVcfBiallelicFiltered.vcf <br/>\n",
    "    $bcftools stats ${outDir}myMergedVcfBiallelicFiltered.vcf.gz > ${outDir}statistics.txt\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [x] Independent Variant Filtering\n",
    "- [x] Multi-sample VCF Generation\n",
    "- [x] Joint Variant Calling/Backfill/Squaring\n",
    "- [x] Split multiallelic variants\n",
    "- [x] Joint Variant Filtering\n",
    "- [ ] <span style=\"color:blue\">**Annotation**</span>\n",
    "\n",
    "<br/>\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Useful Resources and Tips </b><br/><br/><b>Transition and Tranversion Ratio Resources</b><br/>\n",
    "* <i>Wang, J., Raskin, L., Samuels, D. C., Shyr, Y., & Guo, Y. (2015). Genome measures used for quality control are dependent on gene function and ancestry. Bioinformatics (Oxford, England), 31(3), 318–323. <a href=\"https://doi.org/10.1093/bioinformatics/btu668\">https://doi.org/10.1093/bioinformatics/btu668</a></i><br/>\n",
    "* This <a href=\"https://www.biostars.org/p/4751/\">Biostars Ti/Tv Blog</a> also has a good discussion surrounding this ratio. <br/>\n",
    "* <a href=\"https://bioinformatics.stackexchange.com/questions/4362/why-ti-tv-ratio\">Stack Exchange</a> also has a good discussion and mathmatical representation of this ratio as well.\n",
    "</div>\n",
    "<br/>\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painful-stomach",
   "metadata": {},
   "source": [
    "## STEP10: Annotation Using Variant Effect Predictor and/or Annovar\n",
    "Typically for human genomes, I use Annovar to annotate variants since the databases are very well curated.  However, Variant Effect Predictor (VEP) works well for other organisms that are non-human.  Annovar does support some mouse databases, but typically it works best for human genomes.\n",
    "\n",
    "### Commands\n",
    "\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [x] Independent Variant Filtering\n",
    "- [x] Multi-sample VCF Generation\n",
    "- [x] Joint Variant Calling/Backfill/Squaring\n",
    "- [x] Split multiallelic variants\n",
    "- [x] Joint Variant Filtering\n",
    "- [x] Annotation\n",
    "\n",
    "<br/>\n",
    "<hr style=\"border:3px solid orange\"> </hr>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
