{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "formed-marshall",
   "metadata": {},
   "source": [
    "# Whole Exome Sequencing Variant Calling\n",
    "\n",
    "This tutorial will provide an example pipeline for calling variants in whole exome sequencing (WES) data across multiple samples, assuming the samples are paired-end. It is strongly recommended that these step be run on a large compute resource such a high-performance compute cluster (HPC) or a cloud resource.  For HPC, these commands should be wrapped inside a shell script with the HPC scheduling language installed on the cluster.  I will try to denote which ones are critical for the cluster, versus which ones can be run locally.\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "august-treat",
   "metadata": {},
   "source": [
    "## STEP0: Fastq Raw Quality Check\n",
    "\n",
    "Before any sample processing can begin, it is critical to know the initial state of the raw sequences. This is typically delivered from the sequencer in fastq format (.fastq/.fq) or gzipped (fastq.gz/fq.gz) so the data is compressed to decrease file size and storage use. \n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)|\n",
    "| Dependencies             | Perl, Java + JDK |\n",
    "| Tested on                | FastQC==v0.11.9, Ubuntu 18.04.4 LTS (Bionic Beaver)| \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | html report and compressed images |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each fastq file_**</span>, you will want to run the following command:\n",
    "<br/>\n",
    "\n",
    "<p style=\"background:white\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">fastqc myFile.fastq.gz -o /path/to/my/output/directory\n",
    "</code>\n",
    "</p>\n",
    "    \n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &ltinsert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt\n",
    "    \n",
    "    # dependency of fastqc\n",
    "    export PERL5LIB=/path/to/perl\n",
    "    \n",
    "    \n",
    "    # declare and define variables\n",
    "    fastqc=\"/path/to/fastqc\"\n",
    "    listOfFastqs=\"/path/to/file/with/fastq/pathnames.txt\"\n",
    "    outDir=\"/path/to/output/directory/\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP0_fastqc.parallel.log --resume\"\n",
    "    \n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles < $listOfFastqs\n",
    "    \n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $fastqc {} -o $outDir\" ::: ${allFiles[@]}\n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "\n",
    "For each fastq file an html report and a zipped directory will be produced.  The zipped directory contains the same information ast the html report; the difference being that the zipped directory contains all the report images as image files that can be embedded into documents/presentations/etc...\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [ ] <span style=\"color:blue\">**Sequence trimming and clean up**</span>\n",
    "- [ ] Fastq Check on Trimmed Fastq\n",
    "- [ ] Alignment/Mapping\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Useful Resources\n",
    "**FastQC**\n",
    "* There is not a formal paper publication on this software, so cite the software as open-source code or webpage.\n",
    "* For [FastQC Docs](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) click on the link.\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "willing-township",
   "metadata": {},
   "source": [
    "## STEP1: Basic quality trimming and adapter removal\n",
    "\n",
    "Based on the output of fastqc, you will want to make sure poor quality bases are trimmed, any adapter contamination as a result of the library prep or sequencing linkers are removed, and any paired reads that may be too short are *removed together.  This is critical for paired-end reads as the order between the forward (R1) and the reverse (R2) must be maintained.*  Reads that are too short will ineviitbaly lead to high multimappers in the alignment phase and can computationally slow down the aligner.\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [cutadapt](https://cutadapt.readthedocs.io/en/stable/)|\n",
    "| Dependencies             | Python3 |\n",
    "| Tested on                | cutadapt==v3.2, Python>=3.6.9, Ubuntu 18.04.4 LTS (Bionic Beaver)| \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "Given cutadapt is a python package, it can be installed using pip:\n",
    "\n",
    "```\n",
    "    pip3 install --user cutadapt==3.2\n",
    "```\n",
    "\n",
    "Assuming FastQC shows Illumina Universal Adapter contamination, we will want to remove those adapter sequences from both reads ends, trim/remove poor quality sequences, and dicard read pairs that may be too short.  One thing to consider, if the sequences were run on the NovaSeq, quality scores are no longer granular!  The NovaSeq bins Phred quality scores to 4 potential bins of either 2, 12, 23, and 37.  Therfore, when selecting a quality cut-off know that if you pick something between the bin values, it will ony remove/trim sequences based on the lower bin.  [Heng Li has great blog](https://lh3.github.io/2017/07/24/on-nonvaseq-base-quality) about this if you are interested.\n",
    "\n",
    "<br/>\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each pair of fastq files_**</span>, you will want to run the following command:\n",
    "\n",
    "<p style=\"background:white\">\n",
    "<code style=\"background:grey;color:black;font-size:16px;margin-left:40px\">cutadapt -a AGATCGGAAGAG -A AGATCGGAAGAG -q 30 --minimum-length=10 --pair-filter=any -o Read1Cleaned_1.fq.gz -p Read2Cleaned_2.fq.gz Read1.fq.gz Read2.fq.gz\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &ltinsert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt\n",
    "    \n",
    "    export PERL5LIB=/path/to/perl\n",
    "    export PYTHONPATH=/path/to/python/site-packages\n",
    "    \n",
    "    # declare and define variables\n",
    "    cutadapt=\"/path/to/cutadapt\"\n",
    "    listOfPEfastqs=\"/path/to/fastqPE/prefixs.txt\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP1_cutadapt.parallel.log --resume\"\n",
    "    \n",
    "    \n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles < $listOfPEfastqs\n",
    "    \n",
    "    \n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $cutadapt -a AGATCGGAAGAG -A AGATCGGAAGAG -q 30 --minimum-length=10 --pair-filter=any -o {}Cleaned_1.fq.gz -p {}Cleaned_2.fq.gz {}1.fq.gz {}2.fq.gz\" ::: ${allFiles[@]} \n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [ ] <span style=\"color:blue\">**Fastq Check on Trimmed Fastq**</span>\n",
    "- [ ] Alignment/Mapping\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Useful Resources\n",
    "**cutadapt Hallmark Paper**\n",
    "* *Martin, M. (2011). Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal, 17(1), pp. 10-12. doi:https://doi.org/10.14806/ej.17.1.200*\n",
    "* For [cutadapt Docs](https://cutadapt.readthedocs.io/en/stable/guide.html) click on the link.\n",
    "        \n",
    "<br/><br/>  \n",
    "        \n",
    "---------------------------------------\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "robust-expert",
   "metadata": {},
   "source": [
    "<br/><br/>\n",
    "\n",
    "## STEP2: Fastq Cleaned Quality Check\n",
    "\n",
    "This is the exact same step as STEP0, with the exception that now you are feeding in the the cleaned fastq files generated in STEP1 that have now been cleaned.  This step is to ensure all cleaning was performed as expected.\n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | local, HPC, or Cloud |\n",
    "| Operating System         | Windows, MacOS, Linux |\n",
    "| Software                 | [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)|\n",
    "| Dependencies             | Perl, Java + JDK |\n",
    "| Tested on                | FastQC==v0.11.9, Ubuntu 18.04.4 LTS (Bionic Beaver) | \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Output File Type         | html report and compressed images |\n",
    "\n",
    "    \n",
    "<br/>\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each fastq file_**</span>, you will want to run the following command:\n",
    "\n",
    "<p style=\"background:white\">&emsp; &emsp; &emsp; &emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">fastqc myCleanedSTEP1File.fastq.gz -o /path/to/my/output/directory/\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &ltinsert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt\n",
    "    \n",
    "    # dependency of fastqc\n",
    "    export PERL5LIB=/path/to/perl\n",
    "    \n",
    "    \n",
    "    # declare and define variables\n",
    "    fastqc=\"/path/to/fastqc\"\n",
    "    listOfFastqs=\"/path/to/file/with/fastq/cleaned/pathnames.txt\"\n",
    "    outDir=\"/path/to/output/directory/\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 5G --delay 0.2 --jobs 5 -u --progress --joblog STEP2_fastqc.parallel.log --resume\"\n",
    "    \n",
    "    # store the prefix list as an array\n",
    "    readarray allFiles < $listOfFastqs\n",
    "    \n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $fastqc {} -o $outDir\" ::: ${allFiles[@]}\n",
    "\n",
    "</code>\n",
    "</p>\n",
    "\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [ ] <span style=\"color:blue\">**Alignment/Mapping**</span>\n",
    "- [ ] Alignment Quality Check\n",
    "- [ ] Duplicate Removal\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Useful Resources\n",
    "**FastQC**\n",
    "* There is not a formal paper publication on this software, so cite the software as open-source code or webpage.\n",
    "* For [FastQC Docs](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) click on the link.\n",
    "\n",
    "<br/><br/>  \n",
    "\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "speaking-engineer",
   "metadata": {},
   "source": [
    "## STEP3: Aligning Reads to Reference Genome\n",
    "\n",
    "This is typically the most computationally expensive part of the analysis and should be performed on a machine with large enough RAM/memory and cores.  Since this is DNA, there is no need to use a splice-aware aligner, therefore a DNA aligner such as bowtie2 or bwa is fine.  \n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [BWA](http://bio-bwa.sourceforge.net/), [Samtools](http://www.htslib.org/) |\n",
    "| Dependencies             |  |\n",
    "| Tested on                | bwa==v0.7.11, samtools==v1.8 | \n",
    "| Input File Type          | raw sequencing fastq file w/extension *.fastq.gz, *.fq.gz, *.fq, or *.fastq |  \n",
    "| Ouput File Type          | alignment file w/extension *.sam or *.bam |\n",
    "\n",
    "\n",
    "<br/>\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each pair of fastq files_**</span>, you will want to run the following command.  This assumes you have 12 cores available for threading -- 6 for alignment and 6 for sorting: \n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">bwa mem -t 6 /path/to/myBWAIndex/filePrefix Read1Cleaned_1.fq.gz Read2Cleaned_1.fq.gz | samtools sort -@6 -o SampleID_trimmed_sorted.bam -\n",
    "</code> \n",
    "</p>\n",
    "<br/>\n",
    "<div class=\"alert alert-block alert-info\"><span style=\"color:black\"><b>Tip:</b> In order to use BWA, you must index your reference genome first.  This only needs to be done one time per reference genome.  This can be accomlished by using <code style=\"background:black;color:white\">bwa index youReferenceGenome.fasta</code>.  For more information, <a href=\"http://bio-bwa.sourceforge.net/bwa.shtml\">look at the bwa documentation here</a>.</span></div>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available fastq files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &ltinsert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt\n",
    "    \n",
    "    module load samtools/1.8 # or declare variable to samtools executable\n",
    "    \n",
    "    # declare and define variables\n",
    "    bwa=\"/path/to/bwa\"\n",
    "    bwaIndexGenome=\"/path/to/bwa_indices/including/prefix\"\n",
    "    listOfPEfastqs=\"/path/to/fastqPE/prefixs.txt\"\n",
    "    #samtools=\"/path/to/samtools/if/needed\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 50G --delay 0.2 --jobs 2 -u --progress --joblog STEP3_WES_bwa_alignment.parallel.log --resume\"\n",
    "    \n",
    "    # store the prefix file list as an array\n",
    "    readarray allFiles < $listOfPEfastqs\n",
    "    \n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    $gnuParallel \"time $bwa mem -t 6 $bwaIndexGenome {}R1.fq.gz {}R2.fq.gz | samtools sort -@6 -o {}trimmed_sorted.bam -\" ::: ${allFiles[@]}\n",
    "    \n",
    "</code>\n",
    "</p>\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [ ] <span style=\"color:blue\">**Alignment Quality Check**</span>\n",
    "- [ ] <span style=\"color:blue\">**Duplicate Removal**</span>\n",
    "- [ ] Independent Variant Calling\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Useful Resources\n",
    "**Burrows-Wheeler Alignment Tool (BWA) Hallmark Paper**\n",
    "* *Li, H., & Durbin, R. (2009). Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics (Oxford, England), 25(14), 1754–1760. https://doi.org/10.1093/bioinformatics/btp324*\n",
    "* For [BWA Docs](http://bio-bwa.sourceforge.net/), click on the link\n",
    "\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "solved-firmware",
   "metadata": {},
   "source": [
    "## STEP4: Collect and Assess Alignment Quality + Remove Duplicates\n",
    "\n",
    "This can easily be performed on a local computer as the memory and storage footprint are relatively low.  This collect alignment information and statistics, insert size metrics, and duplicate metrics.  The duplicate read removal step is critical for WES and variant calling.  In this case, I physically remove the duplicate reads instead of \"marking\" the duplicates, in case the GATK pipeline is not used for variant calling.  All of these metrics collected should be looked at and if there are enough samples, graphically reprensented using your own custom code to ensure the differences can be explained within and between samples.  **It may be the case that after looking at these metrics, alignmnet paramters need to be adjusted and STEP3 needs to be re-run and re-assessed to fit the needs of the project.**\n",
    "\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | Local, HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [Picard](https://broadinstitute.github.io/picard/), [Samtools](http://www.htslib.org/) |\n",
    "| Dependencies             | Java + JDK |\n",
    "| Tested on                | Picard==v2.21.1, samtools==v1.8 | \n",
    "| Input File Type          | alignment file w/extension *.bam or *sam |\n",
    "| Output File Type         | series of txt files and alignment file w/extension *.bam and index *.bai | \n",
    "\n",
    "\n",
    "<br/>\n",
    "\n",
    "For <span style=\"background:yellow;color:black\">**_each alignment file_**</span>, you will want to run the following sets of commands:\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">samtools flagstat alignmentFileName.bam > alignmentFileName.flagstat\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar CollectAlignmentSummaryMetrics I=alignmentFileName.bam O=picard.metrics.txt R=refGenome.fasta\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar CollectInsertSizeMetrics I=alignmentFileName.bam O=insert_size_metrics.txt H=insert_size_histogram.pdf\n",
    "</code> \n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">java -jar Picard.jar MarkDuplicates I=alignmentFileName.bam O=alignmentNoDups.bam M=markDupMetrics.txt REMOVE_DUPLICATES=true\n",
    "</code>\n",
    "\n",
    "<p style=\"background:white\">&emsp;&emsp;\n",
    "<code style=\"background:grey;color:black;font-size:16px\">samtools index alignmentNoDups.bam\n",
    "</code> \n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "Example submission script <span style=\"background:lightgreen;color:black\">**_if using a HPC with a scheduler and parallelizing with GNUparallel_**</span> for all available alignment files:\n",
    "<p style=\"background:white;border:1px solid gray;margin-left:40px\">\n",
    "<code style=\"background:white\">\n",
    "    #!/bin/bash\n",
    "    &ltinsert scheduler resources here -- #BSUB (LSF), #SBATCH (SLURM), #PBS (PBS)&gt\n",
    "    \n",
    "    module load java\n",
    "    module load picard/2.21.1\n",
    "    module load samtools/1.8\n",
    "    export PERL5LIB=/home/bin/perl\n",
    "    \n",
    "    # declare and define variables\n",
    "    #picardJar=\"/path/to/picard/if/needed\"\n",
    "    #samtools=\"/path/to/samtools/if/needed\"\n",
    "    bamPath=\"/path/to/bam/output/location/directory/\"\n",
    "    outDir=\"/path/to/output/results/\"\n",
    "    bamList=\"/path/to/file/with/bamPrefix.txt\"\n",
    "    fastaFile=\"/path/to/reference/genome.fasta\"\n",
    "    \n",
    "    # store the prefix file list as an array\n",
    "    readarray allBamFiles < $bamList\n",
    "    \n",
    "    # run the command and have GNUparallel auto-submit jobs as resources are available\n",
    "    echo \"STEP4: 1/5...flagstats\"\n",
    "    for bamFile in ${bamPath}*.bam;\n",
    "    do\n",
    "      time samtools flagstat ${bamFile} > ${bamFile}.flagstat\n",
    "    done\n",
    "    wait\n",
    "    sleep 60\n",
    "                                           \n",
    "    echo \"STEP4: 2/5...picard alignment metrics\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_alignmentMetrics_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar CollectAlignmentSummaryMetrics I=${bamPath}{}.bam O=${outDir}{}.picard.metrics.txt R=$fastaFile\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60\n",
    "    \n",
    "    echo \"STEP4: 3/5...collect insert size metrics\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_collectInsertMetrics_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar CollectInsertSizeMetrics I=${bamPath}{}.bam O=${outDir}{}insert_size_metrics.txt H=${outDir}{}insert_size_histogram.pdf\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60\n",
    "    \n",
    "    echo \"STEP4: 4/5...mark duplicates\"\n",
    "    gnuParallel=\"/path/to/GNUparallel --tag --memfree 18G --delay 0.2 --jobs 5 -u --progress --joblog STEP6_markDuplicates_picard.parallel.log --resume\"\n",
    "    $gnuParallel \"time java -jar $picardJar MarkDuplicates I=${bamPath}{}.bam O=${outDir}{}noDups.bam M=${outDir}{}markDupMetrics.txt REMOVE_DUPLICATES=true\" ::: ${allBamFiles[@]}\n",
    "    wait\n",
    "    sleep 60\n",
    "    \n",
    "    echo \"STEP4: 5/5...index new bam\"\n",
    "    for newBam in ${outDir}*.bam;\n",
    "    do\n",
    "        time samtools index ${newBam}\n",
    "    done\n",
    "    wait\n",
    "    sleep 60\n",
    "                                           \n",
    "             \n",
    "                                           \n",
    "</code>\n",
    "</p>\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Expected Output\n",
    "\n",
    "<br/><br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [ ] <span style=\"color:blue\">**Independent Variant Calling**</span>\n",
    "- [ ] Independent Variant Filtering\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "\n",
    "### Useful Resources\n",
    "**SamTools Hallmark Paper**\n",
    "* *Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G., Durbin, R., & 1000 Genome Project Data Processing Subgroup (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics (Oxford, England), 25(16), 2078–2079. https://doi.org/10.1093/bioinformatics/btp352*\n",
    "* For [SamTools Docs](http://www.htslib.org/doc/samtools.html), click on the link\n",
    "* [Dave Tang's Blog](https://davetang.org/wiki/tiki-index.php?page=SAMTools) also has a great wiki and tutorial for SamTools\n",
    "\n",
    "**Picard Tools**\n",
    "* For [Picard Docs](https://broadinstitute.github.io/picard/) click on the link\n",
    "* There does not seem to be a formal publication, so please cite their GitHub page located [here](https://github.com/broadinstitute/picard)\n",
    "    \n",
    "<br/><br/>\n",
    "\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pharmaceutical-tracker",
   "metadata": {},
   "source": [
    "## STEP5: Independent Variant Calling per Sample\n",
    "\n",
    "|Specifications            |                      |\n",
    "|:------------------------:|:--------------------:|\n",
    "| Recommended run location | HPC or Cloud |\n",
    "| Operating System         | Linux/Unix-based |\n",
    "| Software                 | [Platypus](https://www.well.ox.ac.uk/research/research-groups/lunter-group/lunter-group/platypus-a-haplotype-based-variant-caller-for-next-generation-sequence-data) |\n",
    "| Dependencies             | Python2.7 |\n",
    "| Tested on                | Platypus==v0.8.1, Python==v2.7.6 | \n",
    "| Input File Type          | alignment file w/extension *.bam or *sam |\n",
    "| Output File Type         | variant call file w/extension .vcf.gz | \n",
    "\n",
    "\n",
    "\n",
    "### Expected Output\n",
    "<br/>\n",
    "\n",
    "### Next Steps Checklist Status\n",
    "- [x] Raw Fastq Check\n",
    "- [x] Sequence trimming and clean up\n",
    "- [x] Fastq Check on Trimmed Fastq\n",
    "- [x] Alignment/Mapping\n",
    "- [x] Alignment Quality Check\n",
    "- [x] Duplicate Removal\n",
    "- [x] Independent Variant Calling\n",
    "- [ ] <span style=\"color:blue\">**Independent Variant Filtering**</span>\n",
    "- [ ] Multi-sample VCF Generation\n",
    "- [ ] Joint Variant Calling/Backfill/Squaring\n",
    "- [ ] Split muliallelic variants\n",
    "- [ ] Joint Variant Filtering\n",
    "- [ ] Annotation\n",
    "\n",
    "<br/>\n",
    "\n",
    "\n",
    "### Useful Resources\n",
    "\n",
    "**Platypus Hallmark Paper**\n",
    "* *Andy Rimmer, Hang Phan, Iain Mathieson, Zamin Iqbal, Stephen R. F. Twigg, WGS500 Consortium, Andrew O. M. Wilkie, Gil McVean, Gerton Lunter. Integrating mapping-, assembly- and haplotype-based approaches for calling variants in clinical sequencing applications. Nature Genetics (2014) doi:10.1038/ng.3036* \n",
    "* For the [Platypus Docs](https://www.rdm.ox.ac.uk/research/lunter-group/lunter-group/platypus-documentation), click on the link\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
